s4-3b1-pc7300: SYSV_FS removed in linux-6.16.x

The SysV filesystem kernel module is no longer in the linux kernel, starting 6.16. This causes an issue with kernel XML generation and poses a documentation issue that should get resolved.

@seal331: Pinging you as it's finally time.

It seems we have another issue though:
```
make -f Makefile.daveb running
make -C /lib/modules/`uname -r`/build M=/home/zeckma/Projects/Documents/LFS/test-instructions/slfs/s4/s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45/sysv
make[1]: Entering directory '/sources/linux/linux-6.16'
make[2]: Entering directory '/home/zeckma/Projects/Documents/LFS/test-instructions/slfs/s4/s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45/sysv'
  CC [M]  ialloc.o
  CC [M]  balloc.o
  CC [M]  inode.o
  CC [M]  itree.o
  CC [M]  file.o
  CC [M]  dir.o
  CC [M]  namei.o
  CC [M]  super.o
namei.c:297:27: error: initialization of ‘struct dentry * (*)(struct mnt_idmap *, struct inode *, struct dentry *, umode_t)’ {aka ‘struct dentry * (*)(struct mnt_idmap *, struct inode *, struct dentry *, short unsigned int)’} from incompatible pointer type ‘int (*)(struct mnt_idmap *, struct inode *, struct dentry *, umode_t)’ {aka ‘int (*)(struct mnt_idmap *, struct inode *, struct dentry *, short unsigned int)’} [-Wincompatible-pointer-types]
  297 |         .mkdir          = sysv_mkdir,
      |                           ^~~~~~~~~~
namei.c:297:27: note: (near initialization for ‘sysv_dir_inode_operations.mkdir’)
namei.c:133:12: note: ‘sysv_mkdir’ declared here
  133 | static int sysv_mkdir(struct mnt_idmap *idmap, struct inode * dir,
      |            ^~~~~~~~~~
```

Yes, I did try using older C standards via `CC` and `CFLAGS`, no dice. Strangely, making the kernel version point at 6.15 also did not work.

--- a/emu/com-emu/s4-3b1-pc7300.xml
+++ b/emu/com-emu/s4-3b1-pc7300.xml
@@ -54,6 +54,12 @@
           <ulink url="&patch-root;/s4-3b1-pc7300/s4-3b1-pc7300-&s4-3b1-pc7300-minor-version;-shared-lib.patch"/>
         </para>
       </listitem>
+      <listitem>
+        <para>
+          Required patch:
+          <ulink url="&patch-root;/s4-3b1-pc7300/s4-3b1-pc7300-&s4-3b1-pc7300-minor-version;-final-updates-from-kernel.patch"/>
+        </para>
+      </listitem>
     </itemizedlist>
 
     <bridgehead renderas="sect3">s4-3b1-pc7300 Dependencies</bridgehead>
@@ -100,6 +106,13 @@
     </para>
 
 <screen><userinput>patch -Np1 -i ../s4-3b1-pc7300-&s4-3b1-pc7300-minor-version;-fix-kernel-module.patch</userinput></screen>
+
+    <para>
+      Apply a second patch to make the kernel module compatible with Linux
+      6.16:
+    </para>
+
+<screen><userinput>patch -Np1 -i ../s4-3b1-pc7300-&s4-3b1-pc7300-minor-version;-final-updates-from-kernel.patch</userinput></screen>
 
     <para>
       Apply a patch to build the supporting library as a shared library instead

--- a/introduction/welcome/changelog.xml
+++ b/introduction/welcome/changelog.xml
@@ -41,6 +41,16 @@
     -->
 
     <listitem>
+      <para>August 3nd, 2025</para>
+      <itemizedlist>
+        <listitem>
+          <para>[seal331] - s4-3b1-pc7300: Unbreak compile with Linux 6.16.
+          Fixes <ulink url="&slfs-issues;/245">#245</ulink>.</para>
+        </listitem>
+      </itemizedlist>
+    </listitem>
+
+    <listitem>
       <para>August 2nd, 2025</para>
       <itemizedlist>
         <listitem>

--- a/patches/s4-3b1-pc7300/s4-3b1-pc7300-efe55b-final-updates-from-kernel.patch
+++ b/patches/s4-3b1-pc7300/s4-3b1-pc7300-efe55b-final-updates-from-kernel.patch
@@ -1,0 +1,128 @@
+diff -Naur s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45-clean/sysv/namei.c s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45/sysv/namei.c
+--- s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45-clean/sysv/namei.c	2025-08-03 09:20:52.709072668 +0300
++++ s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45/sysv/namei.c	2025-08-03 09:40:25.455831023 +0300
+@@ -130,7 +130,7 @@
+ 	return add_nondir(dentry, inode);
+ }
+ 
+-static int sysv_mkdir(struct mnt_idmap *idmap, struct inode * dir,
++static struct dentry *sysv_mkdir(struct mnt_idmap *idmap, struct inode * dir,
+ 		      struct dentry *dentry, umode_t mode)
+ {
+ 	struct inode * inode;
+@@ -157,7 +157,7 @@
+ 
+         d_instantiate(dentry, inode);
+ out:
+-	return err;
++	return ERR_PTR(err);
+ 
+ out_fail:
+ 	inode_dec_link_count(inode);
+diff -Naur s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45-clean/sysv/super.c s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45/sysv/super.c
+--- s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45-clean/sysv/super.c	2025-08-03 09:20:52.709173149 +0300
++++ s4-3b1-pc7300-efe55b8cc917e3e27bcae9786c5b5dec0469be45/sysv/super.c	2025-08-03 09:42:32.595818909 +0300
+@@ -24,6 +24,7 @@
+ #include <linux/init.h>
+ #include <linux/slab.h>
+ #include <linux/buffer_head.h>
++#include <linux/fs_context.h>
+ #include "sysv.h"
+ 
+ /*
+@@ -410,12 +411,13 @@
+ 	return 1;
+ }
+ 
+-static int sysv_fill_super(struct super_block *sb, void *data, int silent)
++static int sysv_fill_super(struct super_block *sb, struct fs_context *fc)
+ {
+ 	struct buffer_head *bh1, *bh = NULL;
+ 	struct sysv_sb_info *sbi;
+ 	unsigned long blocknr;
+ 	int size = 0, i;
++	int silent = fc->sb_flags & SB_SILENT;
+ 	
+ 	BUILD_BUG_ON(1024 != sizeof (struct xenix_super_block));
+ 	BUILD_BUG_ON(512 != sizeof (struct sysv4_super_block));
+@@ -532,10 +534,11 @@
+ 	return 1;
+ }
+ 
+-static int v7_fill_super(struct super_block *sb, void *data, int silent)
++static int v7_fill_super(struct super_block *sb, struct fs_context *fc)
+ {
+ 	struct sysv_sb_info *sbi;
+ 	struct buffer_head *bh;
++	int silent = fc->sb_flags & SB_SILENT;
+ 
+ 	if (440 != sizeof (struct v7_super_block))
+ 		panic("V7 FS: bad super-block size");
+@@ -589,33 +592,51 @@
+ 
+ /* Every kernel module contains stuff like this. */
+ 
+-static struct dentry *sysv_mount(struct file_system_type *fs_type,
+-	int flags, const char *dev_name, void *data)
++static int sysv_get_tree(struct fs_context *fc)
+ {
+-	return mount_bdev(fs_type, flags, dev_name, data, sysv_fill_super);
++	return get_tree_bdev(fc, sysv_fill_super);
+ }
+ 
+-static struct dentry *v7_mount(struct file_system_type *fs_type,
+-	int flags, const char *dev_name, void *data)
++static int v7_get_tree(struct fs_context *fc)
+ {
+-	return mount_bdev(fs_type, flags, dev_name, data, v7_fill_super);
++	return get_tree_bdev(fc, v7_fill_super);
++}
++
++static const struct fs_context_operations sysv_context_ops = {
++	.get_tree	= sysv_get_tree,
++};
++
++static const struct fs_context_operations v7_context_ops = {
++	.get_tree	= v7_get_tree,
++};
++
++static int sysv_init_fs_context(struct fs_context *fc)
++{
++	fc->ops = &sysv_context_ops;
++	return 0;
++}
++
++static int v7_init_fs_context(struct fs_context *fc)
++{
++	fc->ops = &v7_context_ops;
++	return 0;
+ }
+ 
+ static struct file_system_type sysv_fs_type = {
+-	.owner		= THIS_MODULE,
+-	.name		= "sysv",
+-	.mount		= sysv_mount,
+-	.kill_sb	= kill_block_super,
+-	.fs_flags	= FS_REQUIRES_DEV,
++	.owner			= THIS_MODULE,
++	.name			= "sysv",
++	.kill_sb		= kill_block_super,
++	.fs_flags		= FS_REQUIRES_DEV,
++	.init_fs_context	= sysv_init_fs_context,
+ };
+ MODULE_ALIAS_FS("sysv");
+ 
+ static struct file_system_type v7_fs_type = {
+-	.owner		= THIS_MODULE,
+-	.name		= "v7",
+-	.mount		= v7_mount,
+-	.kill_sb	= kill_block_super,
+-	.fs_flags	= FS_REQUIRES_DEV,
++	.owner			= THIS_MODULE,
++	.name			= "v7",
++	.kill_sb		= kill_block_super,
++	.fs_flags		= FS_REQUIRES_DEV,
++	.init_fs_context	= v7_init_fs_context,
+ };
+ MODULE_ALIAS_FS("v7");
+ MODULE_ALIAS("v7");

